Bài 1.1 



a. Từ quan trọng nhất và lý do:
Fill: Theo em, từ quan trọng nhất là “learn” (học) vì đây là điểm khác biệt lớn nhất giữa Machine Learning và lập trình thông thường. Máy tính không chỉ làm theo lệnh cố định mà còn có thể cải thiện kết quả dựa trên dữ liệu và kinh nghiệm.


b. Định nghĩa ML theo tôi:
Fill: Machine Learning là phương pháp giúp máy tính học từ dữ liệu và tự cải thiện khả năng thực hiện một nhiệm vụ mà không cần được lập trình chi tiết cho từng trường hợp.


Bài 1.2 

| Ứng dụng                              | Task (T)                            | Experience (E)                                        | Performance (P)                                    |
| ------------------------------------- | ----------------------------------- | ----------------------------------------------------- | -------------------------------------------------- |
| Spam Filter                           | Phân loại email spam/ham            | Dữ liệu email đã được gắn nhãn spam hoặc không spam   | Tỷ lệ phân loại đúng email                         |
| Hệ thống gợi ý phim Netflix           | Gợi ý phim phù hợp cho người dùng   | Lịch sử xem, đánh giá và tìm kiếm phim của người dùng | Mức độ người dùng nhấn xem hoặc hài lòng với gợi ý |
| Ứng dụng tự chọn: Nhận diện khuôn mặt | Nhận diện đúng khuôn mặt người dùng | Hình ảnh khuôn mặt đã được lưu và huấn luyện trước    | Độ chính xác khi nhận diện khuôn mặt               |


Bài 1.3 

Fill: Em sẽ chọn Machine Learning vì nhiệt độ phụ thuộc vào rất nhiều yếu tố như độ ẩm, áp suất, gió, mùa và dữ liệu thời tiết trong quá khứ. Những quy luật này khá phức tạp nên khó viết bằng các câu lệnh cố định của Traditional Programming. ML có thể học từ dữ liệu thời tiết trước đó để tìm ra quy luật và dự đoán chính xác hơn.


Bài 2.1 

| Tiêu chí                         | Traditional Programming        | Machine Learning                             |
| -------------------------------- | ------------------------------ | -------------------------------------------- |
| Cách giải quyết vấn đề           | Người viết luật/rules thủ công | Máy tự học quy luật từ dữ liệu               |
| Khi bài toán thay đổi            | Phải sửa lại code/rules        | Cần thêm data, retrain                       |
| Dataset lớn, phức tạp            | Khó — rules bùng nổ            | Xử lý tốt hơn nhờ học từ dữ liệu             |
| Explainability (giải thích được) | Cao — ta biết rule nào fired   | Thấp hơn — khó biết model quyết định thế nào |
| Thời gian xây dựng ban đầu       | Nhanh với bài toán đơn giản    | Cần nhiều data + training                    |


Bìa 2.2 

| Ứng dụng                        | Nhóm (Vision / NLP / Analytics / Other) | Supervised hay Unsupervised? |
| ------------------------------- | --------------------------------------- | ---------------------------- |
| Face unlock trên điện thoại     | Vision                                  | Supervised                   |
| Google Translate                | NLP                                     | Supervised                   |
| Netflix gợi ý phim              | Analytics                               | Unsupervised                 |
| Phát hiện gian lận thẻ tín dụng | Analytics                               | Supervised                   |
| ChatGPT                         | NLP                                     | Supervised                   |
| Xe tự lái                       | Vision                                  | Supervised                   |
| Spotify Discover Weekly         | Analytics                               | Unsupervised                 |
| Dự đoán giá cổ phiếu            | Analytics                               | Supervised                   |


Bài 2.3 

Fill: a. Nguyên nhân gốc rễ của vấn đề này là gì?
Nguyên nhân chính là dữ liệu huấn luyện bị thiên lệch (biased data). Hệ thống học từ CV lịch sử mà phần lớn ứng viên được tuyển là nam giới, nên model hiểu sai rằng nam giới “tốt hơn”. ML chỉ học theo dữ liệu đã cho nên nếu dữ liệu có định kiến thì model cũng sẽ học theo định kiến đó.

Fill: b. Làm thế nào để phát hiện và ngăn chặn bias như vậy?
Có thể phát hiện bias bằng cách kiểm tra kết quả dự đoán giữa các nhóm khác nhau như nam và nữ để xem hệ thống có đối xử công bằng không. Để ngăn chặn bias cần:

Dùng dataset đa dạng và cân bằng hơn.
Loại bỏ các thông tin nhạy cảm như giới tính.
Thường xuyên kiểm tra và đánh giá fairness của model.
Có con người giám sát thay vì để AI quyết định hoàn toàn.

Fill: c. ML có nên được dùng để quyết định tuyển dụng không? Lập luận của bạn?
Theo em, ML có thể hỗ trợ tuyển dụng nhưng không nên quyết định hoàn toàn. AI có thể giúp lọc CV nhanh và tiết kiệm thời gian, nhưng vẫn có nguy cơ bias hoặc đánh giá sai con người. Quyết định cuối cùng nên do con người thực hiện để đảm bảo công bằng và xem xét nhiều yếu tố mà AI khó hiểu hết.

Bài 3.1 

| Loại Learning   | Đặc điểm chính                              | Data có label không?  | Ví dụ ứng dụng                               | Thuật toán tiêu biểu                  |
| --------------- | ------------------------------------------- | --------------------- | -------------------------------------------- | ------------------------------------- |
| Supervised      | Học từ dữ liệu có đáp án đúng (label)       | Có                    | Phân loại spam email, dự đoán giá nhà        | Linear Regression, SVM, Decision Tree |
| Unsupervised    | Không có teacher, tự tìm patterns           | Không                 | Customer segmentation, clustering khách hàng | K-Means, PCA                          |
| Semi-supervised | Kết hợp dữ liệu có label và không có label  | Một phần              | Phân loại ảnh với ít dữ liệu được gắn nhãn   | Self-training, Semi-supervised SVM    |
| Reinforcement   | Agent học bằng cách thử và nhận reward/phạt | Không (chỉ có reward) | Game AI, xe tự lái                           | Q-Learning, Deep Q Network (DQN)      |


Bài 3.2 

| Bài toán                                          | C hay R? | Lý do (output là gì?)                              |
| ------------------------------------------------- | -------- | -------------------------------------------------- |
| Dự đoán giá nhà (triệu đồng)                      | R        | Output là một giá trị số liên tục (giá tiền)       |
| Email: spam hay không?                            | C        | Output là các nhãn phân loại: spam hoặc không spam |
| Dự đoán điểm thi của sinh viên (0-10)             | R        | Output là một giá trị số điểm                      |
| Chẩn đoán bệnh: ung thư / bình thường / lành tính | C        | Output là các nhóm/nhãn bệnh                       |
| Dự đoán số lượng hàng bán ngày mai                | R        | Output là số lượng sản phẩm bán ra                 |
| Nhận diện chữ viết tay: chữ số 0-9                | C        | Output là nhãn chữ số từ 0 đến 9                   |


Bài 3.3 

| # | Bài toán thực tế                                                          | Loại learning   | Lý do nhóm chọn                                                 |
| - | ------------------------------------------------------------------------- | --------------- | --------------------------------------------------------------- |
| 1 | Robot học chơi cờ vua — thắng/thua là reward                              | Reinforcement   | Robot học qua thử nghiệm và nhận reward khi thắng               |
| 2 | Phân nhóm 10 triệu ảnh sản phẩm không có nhãn                             | Unsupervised    | Dữ liệu không có label nên model tự tìm nhóm giống nhau         |
| 3 | Dự đoán SV nào có nguy cơ bỏ học dựa trên điểm số lịch sử                 | Supervised      | Có dữ liệu quá khứ và label về việc bỏ học hay không            |
| 4 | Nhận diện tiếng nói: có 1000 giờ audio có label + 100,000 giờ không label | Semi-supervised | Kết hợp dữ liệu có label và không có label                      |
| 5 | Hệ thống gợi ý sản phẩm dựa trên lịch sử mua hàng                         | Unsupervised    | Hệ thống tìm pattern và sở thích người dùng từ dữ liệu mua hàng |
| 6 | Phát hiện bất thường trong dữ liệu server (log files)                     | Unsupervised    | Khó có label đầy đủ nên model tìm điểm bất thường trong dữ liệu |
| 7 | Dịch tự động anh-việt với dataset 1M cặp câu                              | Supervised      | Có dữ liệu input-output rõ ràng (câu Anh và câu Việt tương ứng) |
| 8 | AlphaGo học cách đánh cờ vây                                              | Reinforcement   | AI học qua chơi nhiều ván và nhận reward khi chiến thắng        |


Bài 4.1 

| Mô tả tình huống                                                                       | Data Problem           |
| -------------------------------------------------------------------------------------- | ---------------------- |
| Model dự đoán lương nhưng data training chỉ có nhân viên văn phòng, không có công nhân | Nonrepresentative Data |
| Dataset chứa nhiều outlier vì có lỗi nhập liệu (ví dụ: tuổi = 999)                     | Poor Quality           |
| Chỉ có 50 ảnh X-quang bệnh hiếm gặp để train model chẩn đoán                           | Insufficient Quantity  |
| Feature “mã màu sản phẩm” không liên quan gì đến việc khách hàng có mua không          | Irrelevant Features    |
| Model nhận dạng giọng nói không nhận ra giọng miền Nam vì train toàn giọng Bắc         | Sampling Bias          |


Bài 4.2

| Scenario | Train Acc | Test Acc | Chẩn đoán        | Giải pháp bạn đề xuất                                  |
| -------- | --------- | -------- | ---------------- | ------------------------------------------------------ |
| A        | 99%       | 54%      | Overfitting      | Thêm data, dùng regularization, giảm độ phức tạp model |
| B        | 62%       | 60%      | Underfitting     | Dùng model mạnh hơn, train lâu hơn, thêm feature       |
| C        | 88%       | 86%      | Good Fit         | Model hoạt động tốt, chỉ cần fine-tune nhẹ nếu cần     |
| D        | 72%       | 70%      | Hơi Underfitting | Cải thiện feature hoặc tăng độ phức tạp model          |
| E        | 95%       | 94%      | Very Good Fit    | Kết quả rất tốt, có thể giữ nguyên model               |


Bài 4.3 

In [2]:
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split

# Tạo dataset
X, y = make_classification(
    n_samples=200,
    n_features=10,
    random_state=42
)

# Chia train/test
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.3,
    random_state=42
)

print("=== Quan sát Overfitting ===\n")

# Thử nhiều độ sâu khác nhau
for depth in [1, 2, 3, 5, 10, None]:

    # Tạo model
    model = DecisionTreeClassifier(
        max_depth=depth,
        random_state=42
    )

    # Train model
    model.fit(X_train, y_train)

    # Accuracy
    train_acc = model.score(X_train, y_train)
    test_acc = model.score(X_test, y_test)

    # In kết quả
    print(f"depth={depth}")
    print(f"Train Accuracy = {train_acc:.3f}")
    print(f"Test Accuracy  = {test_acc:.3f}")
    print("-" * 35)

=== Quan sát Overfitting ===

depth=1
Train Accuracy = 0.900
Test Accuracy  = 0.850
-----------------------------------
depth=2
Train Accuracy = 0.921
Test Accuracy  = 0.850
-----------------------------------
depth=3
Train Accuracy = 0.943
Test Accuracy  = 0.817
-----------------------------------
depth=5
Train Accuracy = 0.964
Test Accuracy  = 0.800
-----------------------------------
depth=10
Train Accuracy = 1.000
Test Accuracy  = 0.750
-----------------------------------
depth=None
Train Accuracy = 1.000
Test Accuracy  = 0.750
-----------------------------------


Fill: Quan sát: Khi depth tăng, train accuracy và test accuracy thay đổi như thế nào?
Khi depth tăng, train accuracy thường tăng liên tục và có thể đạt gần 100%. Test accuracy lúc đầu tăng nhưng sau đó giảm hoặc không tăng nữa vì model bắt đầu học quá chi tiết dữ liệu train.

Fill: Depth nào gây overfitting rõ nhất? Bạn nhận ra bằng dấu hiệu gì?
Depth = None hoặc depth rất lớn như 10 thường gây overfitting rõ nhất. Dấu hiệu là train accuracy rất cao (gần 100%) nhưng test accuracy thấp hơn nhiều, chứng tỏ model nhớ dữ liệu train nhưng tổng quát kém.

Fill: Bạn sẽ chọn depth nào để deploy? Tại sao?
Tôi sẽ chọn depth khoảng 3 hoặc 5 vì train accuracy và test accuracy cân bằng hơn, chênh lệch nhỏ và model có khả năng tổng quát tốt trên dữ liệu mới.

Bài 5.1 

| Split          | Số lượng    | Mục đích (viết bằng lời của bạn)                                            |
| -------------- | ----------- | --------------------------------------------------------------------------- |
| Training Set   | 7,000 (70%) | Dùng để train model và học từ dữ liệu                                       |
| Validation Set | 1,500 (15%) | Dùng để tune hyperparameter và chọn model tốt nhất                          |
| Test Set       | 1,500 (15%) | Dùng để đánh giá cuối cùng xem model hoạt động tốt thế nào trên dữ liệu mới |


Fill: Tại sao KHÔNG được nhìn vào Test Set khi đang tune model?
Vì nếu dùng Test Set để tune model thì model sẽ vô tình “học” luôn dữ liệu test, làm kết quả đánh giá không còn khách quan nữa. Khi đó accuracy trên test sẽ cao giả tạo và không phản ánh đúng khả năng hoạt động thực tế của model trên dữ liệu mới.

Bài 5.2

Fill: Sơ đồ K=5 Cross-Validation (vẽ hoặc mô tả):

Dataset được chia thành 5 phần bằng nhau:

Fold 1 → Validation, Fold 2-5 → Training
Fold 2 → Validation, Fold 1-3-4-5 → Training
Fold 3 → Validation, Fold 1-2-4-5 → Training
Fold 4 → Validation, Fold 1-2-3-5 → Training
Fold 5 → Validation, Fold 1-2-3-4 → Training

Fill: Kết quả cuối cùng của K-Fold là gì? Tại sao tốt hơn dùng 1 validation set?

Kết quả cuối cùng là trung bình accuracy (hoặc metric khác) của cả 5 lần train/validation.

K-Fold tốt hơn dùng 1 validation set vì:

Tất cả dữ liệu đều được dùng để validation ít nhất một lần.
Giảm phụ thuộc vào cách chia dữ liệu ngẫu nhiên.
Đánh giá model ổn định và đáng tin cậy hơn.

Bài 5.3

| Tiêu chí                   | Statistical Modeling                                 | Machine Learning                       |
| -------------------------- | ---------------------------------------------------- | -------------------------------------- |
| Mục tiêu chính             | Hiểu mối quan hệ giữa các biến và giải thích dữ liệu | Dự đoán kết quả chính xác nhất có thể  |
| Câu hỏi tiêu biểu          | “X ảnh hưởng Y thế nào?”                             | “Làm sao dự đoán Y chính xác nhất?”    |
| Dữ liệu cần thiết          | Thường cần dữ liệu nhỏ hoặc vừa, có cấu trúc rõ ràng | Nhiều data, high-dimensional OK        |
| Kết quả có thể giải thích? | Cao, dễ giải thích                                   | Thấp hơn, nhiều model khó giải thích   |
| Ví dụ công cụ              | Linear Regression, ANOVA                             | Random Forest, Neural Network, XGBoost |


Fill: 

a. “Thu nhập có ảnh hưởng đến tỉ lệ mua hàng không?”
→ Phù hợp hơn với Statistical Modeling vì mục tiêu là tìm mối quan hệ và mức độ ảnh hưởng giữa các biến.

b. “Dự đoán khách hàng nào có khả năng mua cao nhất”
→ Phù hợp hơn với Machine Learning vì mục tiêu chính là dự đoán chính xác hành vi khách hàng.

c. “Yếu tố nào giúp sinh viên tốt nghiệp đúng hạn?”
→ Phù hợp với cả hai:

Statistical Modeling để phân tích yếu tố ảnh hưởng.
Machine Learning để dự đoán sinh viên nào có nguy cơ tốt nghiệp trễ.

Bài 6.1

Fill: Bài toán tôi muốn giải quyết bằng ML:
Dự đoán sinh viên có nguy cơ bỏ học dựa trên điểm số, tần suất đi học và hoạt động học tập.

Fill: Tại sao bài toán này quan trọng? (impact là gì?)
Bài toán này quan trọng vì giúp nhà trường phát hiện sớm sinh viên đang gặp khó khăn để hỗ trợ kịp thời. Điều này có thể giảm tỷ lệ bỏ học, cải thiện kết quả học tập và giúp sinh viên có cơ hội tốt nghiệp đúng hạn.

Bài 6.2 

| Thành phần                            | Mô tả chi tiết cho dự án của bạn                                                                        |
| ------------------------------------- | ------------------------------------------------------------------------------------------------------- |
| Task (T) — Bài toán cụ thể            | Dự đoán sinh viên có nguy cơ bỏ học dựa trên điểm số, số buổi vắng và hoạt động học tập                 |
| Experience (E) — Data sẽ dùng         | Dữ liệu sinh viên gồm điểm số, tỷ lệ chuyên cần, lịch sử học tập và trạng thái bỏ học của các năm trước |
| Performance (P) — Đo lường thành công | Accuracy hoặc F1-score cao (ví dụ > 85%) trong việc dự đoán đúng sinh viên có nguy cơ bỏ học            |


| Câu hỏi phân tích                                  | Câu trả lời                                                                                                                          |
| -------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------- |
| Loại learning nào phù hợp?                         | Supervised Learning — vì có dữ liệu quá khứ và label rõ ràng (sinh viên bỏ học hay không)                                                   |
| Classification hay Regression?                     | Classification — output là “có nguy cơ bỏ học” hoặc “không có nguy cơ”                                                                      |
| Data lấy từ đâu?                                   | Kaggle hoặc dữ liệu thu thập từ hệ thống quản lý sinh viên của trường                                                                       |
| Thách thức lớn nhất bạn dự đoán là gì?             | Data quality — dữ liệu có thể thiếu, không đầy đủ hoặc không đồng nhất                                                                      |
| Nếu model đạt 80% accuracy, bạn có hài lòng không? | Phụ thuộc vào mức độ quan trọng của bài toán và dữ liệu. Nếu giúp phát hiện sớm phần lớn sinh viên có nguy cơ thì 80% có thể chấp nhận được |
